### PR TITLE
[BEAM-3219] DataflowRunner: delegate @Setup and @Teardown in stateful ParDo

### PR DESCRIPTION
**Pull request by kennknowles**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
Jenkins configs were broken when this ran.

**Comment by kennknowles** [comment]
run dataflow validatesrunner

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java [hunk 1]
```diff
@@ -204,25 +224,63 @@ public void after() {
...
-        .apply(ParDo.of(fn));
-    try {
-      p.run();
-      fail("Pipeline should have failed with an exception");
-    } catch (Exception e) {
-      assertThat(
-          "Function should have been torn down after exception",
-          ExceptionThrowingOldFn.teardownCalled.get(),
-          is(true));
+  private static class CallSequenceEnforcingStatefulFn<K, V> extends DoFn<KV<K, V>, KV<K, V>> {
```
Comment (robertwb) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java]: Despite the dangers of inheritance, I think it'd be cleaner to take CallSequenceEnforcingFn as a base than have another copy of the implementation.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java [hunk 2]
```diff
@@ -151,6 +156,21 @@ public void testFnCallSequenceMulti() {
...
   }
 
+  @Test
+  @Category({ValidatesRunner.class, UsesStatefulParDo.class})
+  public void testFnCallSequenceStateful() {
+    PCollectionList.of(p.apply("Impolite", Create.of(KV.of("a", 1), KV.of("b", 2), KV.of("a", 4))))
+        .and(
+            p.apply(
+                "Polite", Create.of(KV.of("b", 3), KV.of("a", 5), KV.of("c", 6), KV.of("c", 7))))
+        .apply(Flatten.<KV<String, Integer>>pCollections())
```
Comment (robertwb) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java]: Is the flatten needed here?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java (hunk 1 above)
Comment (kennknowles) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java]: Done. Confirmed that the test still fails without the fix and succeeds with the fix, just for good measure.

**Comment by kennknowles** [comment]
run dataflow ValidatesRunner

**Pull request merged**